core/app/c/[communitySlug]/unauthorized/pageAuthorizationChecks.tsx

```diff
@@ -0,0 +1,194 @@
+	members: Capabilities.editCommunity,
+	settings: Capabilities.editCommunity,
+	apiTokens: Capabilities.editCommunity,
+	docs: Capabilities.editCommunity,
+	apiDocs: [Capabilities.editCommunity, Capabilities.viewPub],
+} as const satisfies PageAuthorizationCheckMap;
+
+type CapabilitiesForTargetCapabilityArray<
+	C extends Capabilities[],
+	R extends string[] = [],
+> = C extends [...infer Rest extends Capabilities[], infer L extends Capabilities]
+	? CapabilitiesForTargetCapabilityArray<
+			Rest,
+			[...R, IdForCapabilityTarget<CapabilitiesTargetForCapability<L>>]
+		>
+	: R;
+
+type IdForCapabilityTarget<T extends CapabilityTarget> = T extends {
+	type: MembershipType.pub;
+	pubId: PubsId;
+}
+	? PubsId
+	: T extends {
+				type: MembershipType.stage;
+				stageId: StagesId;
+		  }
+		? StagesId
+		: T extends {
+					type: MembershipType.community;
+					communityId: CommunitiesId;
+			  }
+			? CommunitiesId
+			: never;
+
+type Targeted<C extends Capabilities | Capabilities[] | null> = C extends Capabilities
+	? [IdForCapabilityTarget<CapabilitiesTargetForCapability<C>>]
+	: C extends Capabilities[]
+		? CapabilitiesForTargetCapabilityArray<C>
+		: never[];
+
+const isSingleCapability = (
+	capability: Capabilities | Capabilities[] | null
+): capability is Capabilities => {
+	return capability !== null && !Array.isArray(capability);
+};
+
+const constructTargetFromCapability = <T extends Capabilities>(
+	capability: T,
+	targetId: string
+): T extends (typeof formCapabilities)[number]
+	? FormTarget
+	: T extends (typeof pubCapabilities)[number]
+		? PubTarget
+		: T extends (typeof stageCapabilities)[number]
+			? StageTarget
+			: T extends (typeof communityCapabilities)[number]
+				? CommunityTarget
+				: never => {
+	if (formCapabilities.some((c) => c === capability)) {
+		const target = {
+			type: MembershipType.form,
+			formId: targetId as FormsId,
+		} satisfies FormTarget;
+
+		return target as any;
+	} else if (pubCapabilities.some((c) => c === capability)) {
+		const target = {
+			type: MembershipType.pub,
+			pubId: targetId as PubsId,
+		} satisfies PubTarget;
+
+		return target as any;
+	} else if (stageCapabilities.some((c) => c === capability)) {
+		const target = {
+			type: MembershipType.stage,
+			stageId: targetId as StagesId,
+		} satisfies StageTarget;
+
+		return target as any;
+	} else if (communityCapabilities.some((c) => c === capability)) {
+		const target = {
+			type: MembershipType.community,
+			communityId: targetId as CommunitiesId,
+		} satisfies CommunityTarget;
+
+		return target as any;
+	}
+
+	throw new Error(`Invalid capability, ${capability} is not a single capability`);
+};
+
+export const userCanViewPage = cache(
+	async <
+		T extends keyof typeof pageAuthorizationChecks,
+		Capability extends
+			(typeof pageAuthorizationChecks)[T] = (typeof pageAuthorizationChecks)[T],
+		Target extends Targeted<Capability> = Targeted<Capability>,
+	>(
+		page: T,
+		userId: UsersId,
```

Comment:
Since this is always called in a context where we've already retrieved the full user row, it would be nice to make this function accept a `User` here and then check `user.isSuperAdmin` before making the capabilities query. I miss my superadmin privileges in communities I'm not a member of 😄

Reply:
good idea!